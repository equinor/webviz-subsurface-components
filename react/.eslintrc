{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "extends":  [
    "plugin:react/recommended",
    "eslint:recommended",
    "plugin:prettier/recommended"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["@typescript-eslint"],
      "extends": [
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:prettier/recommended"
      ]
    }
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "amd": true,
    "browser": true,
    "es6": true,
    "jasmine": true,
    "node": true
  }
}
