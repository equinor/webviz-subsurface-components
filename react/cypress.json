{
    "baseUrl": "http://localhost:6006/?path=/story/deckglmap--default",
    "video": true
}
