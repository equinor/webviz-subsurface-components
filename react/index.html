<html>
    <head>
        <title>Demo application of webviz-subsurface-components</title>
    </head>
    <body>
        <div id='root'></div>
        <link rel="stylesheet" type="text/css" href="output.css">
        <script src="output.js"></script>
    </body>
</html>
