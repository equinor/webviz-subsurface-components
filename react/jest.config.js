/**
 * The jest tests had an error: SyntaxError: Cannot use import statement outside a module
 * After some investigation, it leads to this page https://github.com/facebook/jest/issues/9395.
 */
module.exports = {
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
            "<rootDir>/__mocks__/fileMock.js",
        "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js",
    },
    transform: {
        "\\.(js|ts|jsx|tsx)$": [
            "babel-jest",
            { configFile: "./config/babel.config.json" },
        ],
    },
    transformIgnorePatterns: ["<rootDir>/node_modules/(?!(@webviz)/)"],
};
