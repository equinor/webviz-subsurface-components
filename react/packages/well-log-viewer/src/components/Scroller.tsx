import React, { Component, ReactNode } from "react";

function getScrollbarSizes(): { vertical: number; horizontal: number } {
    // Creating invisible container
    const outer = document.createElement("div");
    outer.style.visibility = "hidden";
    outer.style.overflow = "scroll"; // forcing scrollbar to appear
    //!!! commented to avoid error TS2339: Property 'msOverflowStyle' does not exist on type 'CSSStyleDeclaration'.
    //!!! outer.style.msOverflowStyle = 'scrollbar'; // needed for WinJS apps
    document.body.appendChild(outer);

    const vertical = outer.offsetWidth - outer.clientWidth;
    const horizontal = outer.offsetHeight - outer.clientHeight;

    // Removing temporary elements from the DOM
    document.body.removeChild(outer);

    return { vertical, horizontal };
}

interface Props {
    /**
     * callback with new scroll positions
     */
    onScroll?: (x: number, y: number) => void;
    children?: ReactNode;
}

class Scroller extends Component<Props> {
    scroller: HTMLDivElement | null; // Outer
    scrollable: HTMLDivElement | null; // Inner
    content: HTMLDivElement | null; // Content over inner
    resizeObserver: ResizeObserver;

    constructor(props: Props) {
        super(props);
        this.scroller = null;
        this.scrollable = null;
        this.content = null;

        this.resizeObserver = new ResizeObserver((entries) => {
            const entry = entries[0];
            if (entry && entry.target) {
                const Width = (entry.target as HTMLElement).offsetWidth;
                const Height = (entry.target as HTMLElement).offsetHeight;

                if (this.content) {
                    const { vertical, horizontal } = getScrollbarSizes();

                    this.content.style.width = Width - vertical + "px";
                    this.content.style.height = Height - horizontal + "px";
                }
            }
        });

        this.onScroll = this.onScroll.bind(this);
    }

    componentDidMount(): void {
        if (this.scroller) this.resizeObserver.observe(this.scroller);
    }
    componentWillUnmount(): void {
        if (this.scroller) this.resizeObserver.unobserve(this.scroller);
    }

    /* current position access functions */
    getScrollX(): number {
        const elOuter = this.scroller;
        if (!elOuter) return 0;
        const scrollWidth = elOuter.scrollWidth - elOuter.clientWidth;
        return scrollWidth ? elOuter.scrollLeft / scrollWidth : 0;
    }
    getScrollY(): number {
        const elOuter = this.scroller;
        if (!elOuter) return 0;
        const scrollHeight = elOuter.scrollHeight - elOuter.clientHeight;
        return scrollHeight ? elOuter.scrollTop / scrollHeight : 0;
    }
    getScrollPos(vertical: boolean | undefined): number {
        return vertical ? this.getScrollY() : this.getScrollX();
    }

    /**
     * callback from HTML element
     */
    onScroll(): void {
        const elOuter = this.scroller;
        if (!elOuter) return;
        // notify parent
        if (this.props.onScroll)
            this.props.onScroll(this.getScrollX(), this.getScrollY());
    }

    /* functions to externally set zoom and scroll position */

    /**
     * @param x value to set the horizontal beginning of visible part of content (fraction)
     * @param y value to set the vertical beginning of visible part of content (fraction)
     * @returns true if visible part is changed
     */
    scrollTo(x: number, y: number): boolean {
        if (x < 0.0) x = 0.0;
        else if (x > 1.0) x = 1.0;
        if (y < 0.0) y = 0.0;
        else if (y > 1.0) y = 1.0;

        const elOuter = this.scroller;
        if (!elOuter) return false;

        const scrollLeft = Math.round(
            x * (elOuter.scrollWidth - elOuter.clientWidth)
        );
        const scrollTop = Math.round(
            y * (elOuter.scrollHeight - elOuter.clientHeight)
        );

        if (
            elOuter.scrollLeft !== scrollLeft ||
            elOuter.scrollTop !== scrollTop
        ) {
            elOuter.scrollTo(scrollLeft, scrollTop);
            return true;
        }
        return false;
    }
    /**
     * @param xZoom set X zoom factor of visible part of content
     * @param yZoom set Y zoom factor of visible part of content
     * @returns true if visible part is changed
     */
    zoom(xZoom: number, yZoom: number): boolean {
        const elOuter = this.scroller;
        if (!elOuter) return false;

        const elInner = this.scrollable;
        if (!elInner) return false;

        const widthInner = Math.round(elOuter.clientWidth * xZoom) + "px";
        const heightInner = Math.round(elOuter.clientHeight * yZoom) + "px";

        if (
            elInner.style.width !== widthInner ||
            elInner.style.height !== heightInner
        ) {
            elInner.style.width = widthInner;
            elInner.style.height = heightInner;

            return true;
        }
        return false;
    }

    render(): JSX.Element {
        return (
            <div
                ref={(el) => (this.scroller = el as HTMLDivElement)}
                style={{ overflow: "scroll", width: "100%", height: "100%" }}
                onScroll={this.onScroll}
            >
                <div ref={(el) => (this.scrollable = el as HTMLDivElement)}>
                    <div
                        ref={(el) => (this.content = el as HTMLDivElement)}
                        style={{ position: "absolute" }}
                    >
                        {this.props.children}
                    </div>
                </div>
            </div>
        );
    }
}

export default Scroller;
