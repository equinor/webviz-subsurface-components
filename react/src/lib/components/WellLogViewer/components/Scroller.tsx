import React, { Component, ReactNode } from "react";

interface Props {
    zoomX?: number;
    zoomY?: number;

    x: number; // fraction
    y: number; // fraction
    onScroll: (x: number, y: number) => void;
}

function getScrollbarSizes(): { vertical: number; horizontal: number } {
    // Creating invisible container
    const outer = document.createElement("div");
    outer.style.visibility = "hidden";
    outer.style.overflow = "scroll"; // forcing scrollbar to appear
    //!!! commented to avoid error TS2339: Property 'msOverflowStyle' does not exist on type 'CSSStyleDeclaration'.
    //!!! outer.style.msOverflowStyle = 'scrollbar'; // needed for WinJS apps
    document.body.appendChild(outer);

    // Creating inner element and placing it in the container
    const inner = document.createElement("div");
    outer.appendChild(inner);

    // Calculating difference between container's full width and the child width
    const vertical = outer.offsetWidth - inner.offsetWidth; // vertical scrollbar
    const horizontal = outer.offsetHeight - inner.offsetHeight; // horizontal scrollbar

    // Removing temporary elements from the DOM
    document.body /*outer.parentNode*/
        .removeChild(outer);

    return {
        vertical,
        horizontal,
    };
}

class Scroller extends Component<Props> {
    scroller: React.RefObject<HTMLInputElement>;
    scrollable: React.RefObject<HTMLInputElement>;

    constructor(props: Props) {
        super(props);

        this.scroller = React.createRef();
        this.scrollable = React.createRef();

        this.onScroll = this.onScroll.bind(this);
    }

    componentDidMount(): void {
        //this.scrollTo(this.props.x, this.props.y);
    }

    componentDidUpdate(prevProps: Props): void {
        if (this.props.x !== prevProps.x || this.props.y !== prevProps.y) {
            this.scrollTo(this.props.x, this.props.y);
        }
    }

    onScroll(): void {
        // callback from HTML element
        const el = this.scroller.current as HTMLElement;
        const scrollTop = el.scrollTop;
        const scrollHeight = el.scrollHeight - el.clientHeight;
        const scrollLeft = el.scrollLeft;
        const scrollWidth = el.scrollWidth - el.clientWidth;
        /*
        console.log(
            "scrollTop=" +
                scrollTop +
                " scrollLeft=" +
                scrollLeft +
                " scrollHeight=" +
                scrollHeight +
                " scrollWidth=" +
                scrollWidth
        );
        */
        // compute fractions
        const x = scrollWidth ? scrollLeft / scrollWidth : 0;
        const y = scrollHeight ? scrollTop / scrollHeight : 0;
        console.log("from scrollbars x=" + x + " y=" + y);

        this.props.onScroll(x, y); // notify parent
        /*
        scrollTo(this.logController, this.props.horizontal ? f2 : f1)
        const posMax = this._scrollPosMax();
        //let pos = this.props.horizontal ? scrollTop: scrollLeft;
        let pos = (this.props.horizontal ? f1 : f2) * posMax;
        console.log("pos=" + pos, "horizontal=" + this.props.horizontal)
        this.scrollTo(pos)
        */
    }

    scrollTo(x: number, y: number): void {
        console.log("scrollTo(" + x + "," + y + ")");
        if (x < 0.0) x = 0.0;
        if (x > 1.0) x = 1.0;
        if (y < 0.0) y = 0.0;
        if (y > 1.0) y = 1.0;

        const { vertical, horizontal } = getScrollbarSizes();
        const el = this.scroller.current as HTMLElement;
        const width = el.getBoundingClientRect().width;
        const height = el.getBoundingClientRect().height;

        const el2 = this.scrollable.current as HTMLElement;
        const width2 = el2.getBoundingClientRect().width + vertical;
        const height2 = el2.getBoundingClientRect().height + horizontal;

        //console.log("el.scroll=", el.scrollLeft, el.scrollTop);
        let left = x * (width2 - width);
        let top = y * (height2 - height);
        left = Math.round(left);
        top = Math.round(top);

        if (el.scrollLeft !== left || el.scrollTop !== top) {
            //console.log(left, top);
            el.scrollLeft = left;
            el.scrollTop = top;
        }
    }

    render(): ReactNode {
        const { vertical, horizontal } = getScrollbarSizes();
        const Width = 1165;
        const Height = 390;

        const width = Width - vertical;
        const height = Height - horizontal;

        const wZ = this.props.zoomX ? this.props.zoomX : 1;
        const hZ = this.props.zoomY ? this.props.zoomY : 1;

        const width2 = width * wZ;
        const height2 = height * hZ;

        //const left = this.props.x * (width2 - width);
        //const top = this.props.y * (height2 - height);
        //console.log("render scrollTo(" + left + "," + top + ")", wZ, hZ);
        //console.log("Scroller render zoom=", wZ, hZ);
        return (
            <div style={{ width: "100%", height: "100%" }}>
                <div
                    style={{
                        overflow: "scroll",
                        width: Width + "px",
                        height: Height + "px",
                    }}
                    ref={this.scroller}
                    onScroll={this.onScroll}
                >
                    <div
                        style={{
                            /*left: -left + "px",
                            top: -top + "px",*/
                            width: width2 + "px",
                            height: height2 + "px",
                        }}
                        ref={this.scrollable}
                    >
                        <div
                            style={{
                                position: "absolute",
                                width: width + "px",
                                height: height + "px",
                            }}
                        >
                            {this.props.children}
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default Scroller;
