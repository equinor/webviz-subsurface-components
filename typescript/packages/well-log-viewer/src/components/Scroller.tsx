import type { ReactNode } from "react";
import React, {
    useRef,
    useEffect,
    useCallback,
    useImperativeHandle,
    forwardRef,
} from "react";

function getScrollbarSizes(): { vertical: number; horizontal: number } {
    // Creating invisible container
    const outer = document.createElement("div");
    outer.style.visibility = "hidden";
    outer.style.overflow = "scroll"; // forcing scrollbar to appear
    //!!! commented to avoid error TS2339: Property 'msOverflowStyle' does not exist on type 'CSSStyleDeclaration'.
    //!!! outer.style.msOverflowStyle = 'scrollbar'; // needed for WinJS apps
    document.body.appendChild(outer);

    const vertical = outer.offsetWidth - outer.clientWidth;
    const horizontal = outer.offsetHeight - outer.clientHeight;

    // Removing temporary elements from the DOM
    document.body.removeChild(outer);

    return { vertical, horizontal };
}

export interface ScrollerProps {
    /**
     * callback with new scroll positions
     */
    onScroll?: (x: number, y: number) => void;
    children?: ReactNode;
}

export interface ScrollerRef {
    getScrollX(): number;
    getScrollY(): number;
    getScrollPos(vertical: boolean | undefined): number;
    scrollTo(x: number, y: number): boolean;
    zoom(xZoom: number, yZoom: number): boolean;
}

const Scroller = forwardRef<ScrollerRef, ScrollerProps>(
    ({ onScroll, children }, ref) => {
        const scrollerRef = useRef<HTMLDivElement>(null);
        const scrollableRef = useRef<HTMLDivElement>(null);
        const contentRef = useRef<HTMLDivElement>(null);
        const resizeObserverRef = useRef<ResizeObserver | null>(null);

        useEffect(() => {
            if (!resizeObserverRef.current) {
                resizeObserverRef.current = new ResizeObserver(
                    (entries: ResizeObserverEntry[]): void => {
                        const entry = entries[0];
                        if (entry && entry.target) {
                            const Width = (entry.target as HTMLElement)
                                .offsetWidth;
                            const Height = (entry.target as HTMLElement)
                                .offsetHeight;

                            if (contentRef.current) {
                                const { vertical, horizontal } =
                                    getScrollbarSizes();

                                contentRef.current.style.width =
                                    Width - vertical + "px";
                                contentRef.current.style.height =
                                    Height - horizontal + "px";
                            }
                        }
                    }
                );
            }

            const scrollerElem = scrollerRef.current;

            if (scrollerElem && resizeObserverRef.current) {
                resizeObserverRef.current.observe(scrollerElem);
            }

            return () => {
                if (scrollerElem && resizeObserverRef.current) {
                    resizeObserverRef.current.unobserve(scrollerElem);
                }
            };
        }, []);

        /* current position access functions */
        const getScrollX = useCallback((): number => {
            const elOuter = scrollerRef.current;
            if (!elOuter) return 0;
            const scrollWidth = elOuter.scrollWidth - elOuter.clientWidth;
            return scrollWidth ? elOuter.scrollLeft / scrollWidth : 0;
        }, []);

        const getScrollY = useCallback((): number => {
            const elOuter = scrollerRef.current;
            if (!elOuter) return 0;
            const scrollHeight = elOuter.scrollHeight - elOuter.clientHeight;
            return scrollHeight ? elOuter.scrollTop / scrollHeight : 0;
        }, []);

        const getScrollPos = useCallback(
            (vertical: boolean | undefined): number => {
                return vertical ? getScrollY() : getScrollX();
            },
            [getScrollX, getScrollY]
        );

        /**
         * callback from HTML element
         */
        const handleScroll = useCallback((): void => {
            const elOuter = scrollerRef.current;
            if (!elOuter) return;
            // notify parent
            onScroll?.(getScrollX(), getScrollY());
        }, [onScroll, getScrollX, getScrollY]);

        /* functions to externally set zoom and scroll position */

        /**
         * @param x value to set the horizontal beginning of visible part of content (fraction)
         * @param y value to set the vertical beginning of visible part of content (fraction)
         * @returns true if visible part is changed
         */
        const scrollTo = useCallback((x: number, y: number): boolean => {
            if (x < 0.0) x = 0.0;
            else if (x > 1.0) x = 1.0;
            if (y < 0.0) y = 0.0;
            else if (y > 1.0) y = 1.0;

            const elOuter = scrollerRef.current;
            if (!elOuter) return false;

            const scrollLeft = Math.round(
                x * (elOuter.scrollWidth - elOuter.clientWidth)
            );
            const scrollTop = Math.round(
                y * (elOuter.scrollHeight - elOuter.clientHeight)
            );

            if (
                elOuter.scrollLeft !== scrollLeft ||
                elOuter.scrollTop !== scrollTop
            ) {
                elOuter.scrollTo(scrollLeft, scrollTop);
                return true;
            }
            return false;
        }, []);

        /**
         * @param xZoom set X zoom factor of visible part of content
         * @param yZoom set Y zoom factor of visible part of content
         * @returns true if visible part is changed
         */
        const zoom = useCallback((xZoom: number, yZoom: number): boolean => {
            const elOuter = scrollerRef.current;
            if (!elOuter) return false;

            const elInner = scrollableRef.current;
            if (!elInner) return false;

            const widthInner = Math.round(elOuter.clientWidth * xZoom) + "px";
            const heightInner = Math.round(elOuter.clientHeight * yZoom) + "px";

            if (
                elInner.style.width !== widthInner ||
                elInner.style.height !== heightInner
            ) {
                elInner.style.width = widthInner;
                elInner.style.height = heightInner;

                return true;
            }
            return false;
        }, []);

        useImperativeHandle(
            ref,
            () => ({
                getScrollX,
                getScrollY,
                getScrollPos,
                scrollTo,
                zoom,
            }),
            [getScrollX, getScrollY, getScrollPos, scrollTo, zoom]
        );

        return (
            <div
                ref={scrollerRef}
                style={{ overflow: "scroll", width: "100%", height: "100%" }}
                onScroll={handleScroll}
            >
                <div ref={scrollableRef}>
                    <div ref={contentRef} style={{ position: "absolute" }}>
                        {children}
                    </div>
                </div>
            </div>
        );
    }
);

Scroller.displayName = "Scroller";

export default Scroller;
