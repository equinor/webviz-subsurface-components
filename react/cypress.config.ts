import { defineConfig } from "cypress";
import getCompareSnapshotsPlugin from "cypress-image-diff-js/dist/plugin";

export default defineConfig({
  component: {
    devServer: {
      framework: "create-react-app",
      bundler: "webpack",
    },
    video:false,
    setupNodeEvents(on, config) {
      getCompareSnapshotsPlugin(on, config),
      on('before:browser:launch', (browser = {}, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.args.push('--start-fullscreen')
      
          return launchOptions
        }
        if (browser.name === 'electron') {
          launchOptions.preferences.fullscreen = true
      
          return launchOptions
        }
      })
    }
  },
});
