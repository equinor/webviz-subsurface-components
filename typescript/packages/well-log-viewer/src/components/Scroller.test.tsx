import { render } from "@testing-library/react";
import "jest-styled-components";
import "@testing-library/jest-dom";
import React from "react";
import Scroller from "./Scroller";

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe("Test scroller", () => {
    it("snapshot test", () => {
        const { container } = render(
            <Scroller onScroll={(x: number, y: number) => [x, y]} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
