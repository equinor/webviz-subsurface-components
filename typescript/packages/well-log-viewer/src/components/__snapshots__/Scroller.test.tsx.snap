// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test scroller snapshot test 1`] = `
<div
  style="overflow: scroll; width: 100%; height: 100%;"
>
  <div>
    <div
      style="position: absolute;"
    />
  </div>
</div>
`;
